File: combination_sum.rb
# 39
def combination_sum(candidates, target)
  result = []
  while candidates.length > 0
    combination_sum_helper(candidates, [candidates[0]], result, target)
    candidates.shift
  end
  result
end

def combination_sum_helper(candidates, solution, result, target)
  sum = solution.reduce(&:+)

  if sum == target
    result << solution.clone
  elsif sum < target
    candidates.each_with_index do |num, i|
      solution << num
      combination_sum_helper(candidates[i..-1], solution, result, target)
      solution.pop
    end
  end
end

File: backtracking/combination_sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
  const solutions = [];
  helper(candidates, 0, [], target, solutions);
  return solutions;
};

function helper(candidates, start, solution, target, solutions) {
  if (sum(solution) === target) {
    solutions.push(solution.slice());
    return;
  } else if (sum(solution) > target) {
    return;
  }

  for (let i = start; i < candidates.length; i += 1) {
    solution.push(candidates[i]);
    helper(candidates, i, solution, target, solutions);
    solution.pop();
  }
}

function sum(arr) {
  let sum = 0;
  arr.forEach((num) => {
    sum += num;
  });
  return sum;
}
